app/src/main/java/com/eveningoutpost/dexdrip/cgm/carelinkfollow/CareLinkFollowService.java

```diff
@@ -0,0 +1,293 @@
+        return Constants.SECOND_IN_MS * gracePeriod;
+    }
+
+    private static long getMissedIntervalMillis() {
+        if(missedPollInterval == 0)
+            return SAMPLE_PERIOD;
+        else
+            return Constants.MINUTE_IN_MS * missedPollInterval;
+    }
+
+    static void scheduleWakeUp() {
+        final BgReading lastBg = BgReading.lastNoSenssor();
+        final long last = lastBg != null ? lastBg.timestamp : 0;
+
+        final long next = anticipateNextWakeUp(JoH.tsl(), last, SAMPLE_PERIOD, getGraceMillis(), getMissedIntervalMillis());
+        wakeup_time = next;
+        UserError.Log.d(TAG, "Anticipate next: " + JoH.dateTimeText(next) + "  last BG timestamp: " + JoH.dateTimeText(last));
+
+        JoH.wakeUpIntent(xdrip.getAppContext(), JoH.msTill(next), WakeLockTrampoline.getPendingIntent(CareLinkFollowService.class, Constants.CARELINK_SERVICE_FAILOVER_ID));
+    }
+
+    public static long anticipateNextWakeUp(long now, final long last, final long period, final long grace, final long missedInterval) {
+
+        long next;
+
+        //recent reading (less then data period) => last + period + grace
+        if((now - last) < period) {
+            next = last + period + grace;
+        }
+        //old reading => anticipated next + grace
+        else{
+            //last expected
+            next = now + ((last - now) % period);
+            //add missed poll interval until future time is reached
+            while(next < now){
+                next += missedInterval;
+            }
+            //add grace
+            next += grace;
+        }
+
+        return  next;
+
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        final PowerManager.WakeLock wl = JoH.getWakeLock("CareLinkFollow-osc", 60000);
+        try {
+
+            UserError.Log.d(TAG, "WAKE UP WAKE UP WAKE UP");
+
+            // Check service should be running
+            if (!shouldServiceRun()) {
+                UserError.Log.d(TAG, "Stopping service due to shouldServiceRun() result");
+                msg("Stopping");
+                stopSelf();
+                return START_NOT_STICKY;
+            }
+            buggySamsungCheck();
+
+            last_wakeup = JoH.tsl();
+
+            // Check current
+            if(gracePeriod == 0)
+                gracePeriod = Pref.getStringToInt("clfollow_grace_period", 30);
+            if(missedPollInterval == 0)
+                missedPollInterval = Pref.getStringToInt("clfollow_missed_poll_interval", 5);
+            lastBg = BgReading.lastNoSenssor();
+            if (lastBg != null) {
+                lastBgTime = lastBg.timestamp;
+            }
+            if (lastBg == null || msSince(lastBg.timestamp) > SAMPLE_PERIOD) {
+                // Get the data
+                if (downloader == null) {
+                    downloader = new CareLinkFollowDownloader(
+                            Pref.getString("clfollow_user", ""),
```

Comment:
How do you ensure that the parameters are not empty ("")?

Reply:
I have taken over this logic from Dexcom share follow service (`ShareFollowService`). Create the downloader with parameters set by the user, which is empty if not set. Then I check the parameters in the downloader (`CareLinkFollowDownloader` `loginDataLooksOkay`) and notify the user in `doEverything`.